### PR TITLE
Add window_size, threshold, composition-based stats, and gap costs options

### PR DESCRIPTION
**Pull request by caleb-easterly**
The four new parameters were added to the Advanced Options drop-down menu. At least one of these parameters were relevant to `blastp`, `blastx`, `tblastx`, `blastn`, `tblastn`, `blastn`, and `rpsblast`, so each of those wrappers was changed slightly. 

I changed the behavior of defaults so that the field is blank if the default should be used (in the case of `integer` parameter type) or a "Use Defaults" is selected (in the case of a `select` parameter). This was to avoid a mix of 0 for default, -1 for default, etc. statements, as 0 is a valid value for some parameters. However, it is a little disconcerting to see a blank field.  

For protein-related searches, I used @jj-umn's excellent list of valid matrix and gap cost combinations. They are combined into a single macro, which was used in `blastp`, `blastx`, and `tblastn`. However, individual `input_gapopen` and `input_gapextend` macros were created for use in `blastn` and `makeprofiledb`. In addition, I improved the documentation for `blastp`.

### CODE REVIEW

File: tools/ncbi_blast_plus/ncbi_blastp_wrapper.xml [hunk 1]
```diff
@@ -17,35 +17,44 @@ blastp
...
-    -matrix $adv_opts.matrix
-    @ADV_FILTER_QUERY@
-    @ADV_MAX_HITS@
-    @ADV_WORD_SIZE@
-    ##Ungapped disabled for now - see comments below
-    ##$adv_opts.ungapped
-    @ADV_ID_LIST_FILTER@
-    @ADV_QCOV_HSP_PERC@
-    $adv_opts.use_sw_tback
+@ADV_MATRIX_GAPCOSTS@
```
Comment (peterjc) [inline tools/ncbi_blast_plus/ncbi_blastp_wrapper.xml]: You've removed the deliberate indentation

File: tools/ncbi_blast_plus/ncbi_blastp_wrapper.xml [hunk 2]
```diff
@@ -17,35 +17,44 @@ blastp
...
-        <param name="blast_type" argument="-task" type="select" display="radio" label="Type of BLAST">
+        <param name="blast_type" argument="-task" type="select" display="radio"
+            label="Type of BLAST"
+            help="See help text for default parameter values for each BLAST type.">
             <option value="blastp">blastp - Traditional BLASTP to compare a protein query to a protein database</option>
-            <option value="blastp-fast">blastp-fast - Use longer words for seeding, faster but less accurate</option>
             <option value="blastp-short">blastp-short - BLASTP optimized for queries shorter than 30 residues</option>
-        </param>
+            <option value="blastp-fast">blastp-fast - Use longer words for seeding, faster but less accurate</option>
+                    </param>
```
Comment (peterjc) [inline tools/ncbi_blast_plus/ncbi_blastp_wrapper.xml]: The indentation of ``</param>`` looks wrong

File: tools/ncbi_blast_plus/ncbi_blastp_wrapper.xml [hunk 3]
```diff
@@ -140,10 +151,54 @@ using the NCBI BLAST+ blastp command line tool.
 
 -----
 
-@OUTPUT_FORMAT@
+**Default parameters for each type of BLAST**
```
Comment (peterjc) [inline tools/ncbi_blast_plus/ncbi_blastp_wrapper.xml]: i don't like the idea of having to maintain a copy of the NCBI BLAST+ documentation values here - the NCBI do change defaults, and it would be all too easy for this text to get out of sync.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 4]
```diff
@@ -258,35 +417,71 @@
...
+        <param argument="-threshold" type="integer" optional="true" min="0"
+            label="Minimum score to add a word to the BLAST lookup table."
+            help="Leave blank for default, which varies based on application."/>
+    </xml>
+
+    <xml name="input_comp_based_stats">
+        <param argument="-comp_based_stats" type="select"
+            label="Composition-based statistics"
+            help="The default value varies based on application and task. Most common default is 2.">
+            <option value="-1">Use default</option>
```
Comment (peterjc) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: How about using a non-integer value for the default, e.g. the empty string or ``"default"`` which would be explicit?

File: tools/ncbi_blast_plus/ncbi_blastp_wrapper.xml (hunk 3 above)
Comment (caleb-easterly) [reply tools/ncbi_blast_plus/ncbi_blastp_wrapper.xml]: Makes sense. Is a link to the command line options (https://www.ncbi.nlm.nih.gov/books/NBK279675/) better? That's what I used to make this table, but it's not clear how often they update that page.

File: tools/ncbi_blast_plus/ncbi_macros.xml (hunk 4 above)
Comment (peterjc) [reply tools/ncbi_blast_plus/ncbi_macros.xml]: Using the blank string you could have the command token macro be consistent with most of the others.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 5]
```diff
@@ -381,11 +589,41 @@
 #end if
     </token>
     <token name="@ADV_WORD_SIZE@">
-#if (str($adv_opts.word_size) and int(str($adv_opts.word_size)) > 0):
+#if str($adv_opts.word_size) != "":
```
Comment (peterjc) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: This can probably be further simplified or made more Pythonic is you put:

```
#if str($adv_opts.word_size):
```

(This is what you've done in the ``@ADV_MATRIX_GAPCOSTS@`` token macro)

**Comment by peterjc** [comment]
I can see why you and @jj-umn have opted to combine matrix and gap costs, but are you happy with preventing the user picking their own gap costs?

I'm OK with using the empty string in the advanced parameters to mean use the defaults (and do not add anything to the command line). From memory older versions of Galaxy did not always respect the optional setting which is likely why I used zero as a sentinel value for defaults.

Are you familiar with ``git rebase`` (rather than using ``git merge``)? i.e. ``git fetch origin`` then ``git rebase origin/master`` assuming you've cloned from my repository as ``origin``.

File: tools/ncbi_blast_plus/ncbi_blastp_wrapper.xml (hunk 3 above)
Comment (peterjc) [reply tools/ncbi_blast_plus/ncbi_blastp_wrapper.xml]: That sounds like a good solution to me 👍

**Comment by caleb-easterly** [comment]
@jj-umn, there was some mention of these being the gap costs that are 'compatible' with each substitution matrix, correct? I'm not sure if I'm remembering that correctly. For me, it seems like the most important thing is that the right defaults pop up for each matrix. 

Another question: how would you feel about moving the `evalue` parameter to under the advanced options? Currently, the default `evalue` parameter doesn't change with the BLAST type, although the defaults for `blastp-short` and `blastp` are 200,000 and 10, respectively (on the web service). 

I just realized that if the defaults (empty strings) are used when the tool is run, there would be no way in Galaxy to find out what parameters were used - even the command line record would not show parameters passed to it. In that way, it's useful to have a less centralized (and less macro-dependent) structure, where one could choose `blastp-short`, for instance, and then have `evalue` populate to 200,000, `word_size` populate to 2, `matrix` to "PAM30", etc. But then we'd have to manually update the defaults (and it seems like that CLI options page doesn't even show the most recent parameters used on the web service - the only way I can find the most recent defaults is by running a query). Thoughts on that?

**Comment by caleb-easterly** [comment]
I'll use rebasing from now on. It makes the history cleaner, yes?

**Comment by nsoranzo** [comment]
Beware that rebases don't send GitHub notifications, so you'll have to add a comment to let us know.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 6]
```diff
@@ -108,18 +111,169 @@
...
+            help="Leave blank for default. See tool help for defaults."/>
+    </xml>
+
     <xml name="input_scoring_matrix">
         <param argument="-matrix" type="select" label="Scoring matrix">
+            <option value="" selected="true">Use default</option>
             <option value="BLOSUM90">BLOSUM90</option>
             <option value="BLOSUM80">BLOSUM80</option>
-            <option value="BLOSUM62" selected="true">BLOSUM62 (default)</option>
+            <option value="BLOSUM62">BLOSUM62 (default)</option>
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: I think it would be better to revert the changes to this macro, since basically the default here is always BLOSUM62 and the empty value breaks the command of `ncbi_makeprofiledb.xml`

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 7]
```diff
@@ -108,18 +111,169 @@
...
             <option value="BLOSUM50">BLOSUM50</option>
             <option value="BLOSUM45">BLOSUM45</option>
             <option value="PAM250">PAM250</option>
             <option value="PAM70">PAM70</option>
             <option value="PAM30">PAM30</option>
         </param>
     </xml>
+
+
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: You can remove one blank line.

**Comment by peterjc** [comment]
@caleb-easterly regarding defaults, the NCBI website does a lot of dynamic option setting based on the input sequence meaning it is often hard to match at the command line or QBLAST API (I've seen this a lot over the years with people calling BLAST via Biopython, gap settings are often part of this).

I think the best route is not to set a default in the wrapper, and let the NCBI command line tool use its own defaults. We capture the command line tool version, so it is possible to determine what BLAST says it would use, and thus reproduce it.

File: tools/ncbi_blast_plus/ncbi_macros.xml (hunk 6 above)
Comment (peterjc) [reply tools/ncbi_blast_plus/ncbi_macros.xml]: Checking ``makeprofiledb``, it too has ``-matrix`` with ``BLOSUM62`` as the default:

```
 -matrix <String, `BLOSUM45', `BLOSUM50', `BLOSUM62', `BLOSUM80', `BLOSUM90',
                  `PAM250', `PAM30', `PAM70'>
   Scoring matrix name
   Default = `BLOSUM62'
```

So alternatively we'd have to update the ``<command>`` to handling this with an in statement, currently it uses:

```
#if str($contain_pssm_scores.contain_pssm_scores_type) == 'no':
    -gapopen $contain_pssm_scores.gapopen
    -gapextend $contain_pssm_scores.gapextend
    -scale $contain_pssm_scores.scale
    -matrix $contain_pssm_scores.matrix
#end if
```

Comment (caleb-easterly) [reply tools/ncbi_blast_plus/ncbi_macros.xml]: The matrix default is almost always BLOSUM62 - the one exception is important in the `blastp` wrapper, because if someone chooses `blastp-short` and then opens advanced options, it looks like the default is BLOSUM62 when it should be PAM30. A small thing, but it comes up fairly often.

Comment (caleb-easterly) [reply tools/ncbi_blast_plus/ncbi_macros.xml]: Ah, I realized that the `input_scoring_matrix` macro is only used in situations where BLOSUM62 is indeed the default, so I reverted it.

**Comment by caleb-easterly** [comment]
The gap cost select list is based on the BLAST website, where the user can't select anything beyond the list of options. I'd be fine with keeping it restricted to the BLAST website options, pending someone needing a non-standard combination. But I'm open to having a more flexible gap cost interface also.

**Comment by peterjc** [comment]
(Replying to an earlier question, yes, ``git rebase`` is useful for a cleaner history)

**Comment by peterjc** [comment]
Regarding:

> Another question: how would you feel about moving the evalue parameter to under the advanced options? Currently, the default evalue parameter doesn't change with the BLAST type, although the defaults for blastp-short and blastp are 200,000 and 10, respectively (on the web service).

Despite e-value being a very poor approach to setting a threshold for when to include hits (since it is database size dependent), it is the single most used BLAST parameter - at least in the papers I read. I think it would be missed if we moved it under the advanced section, and people would complain.

File: tools/ncbi_blast_plus/ncbi_macros.xml (hunk 6 above)
Comment (peterjc) [reply tools/ncbi_blast_plus/ncbi_macros.xml]: If necessary, maybe add a comment to this macro warning about those cases where BLOSUM62 is not the default?

**Comment by peterjc** [comment]
According to ``blastp -help``, also ``blastn -help``, etc

```
 -gapopen <Integer>
   Cost to open a gap
 -gapextend <Integer>
   Cost to extend a gap
```

I would infer that the defaults are determined depending on the other settings (probably the matrix and task in particular), but we'd have to check the source code to confirm that.

So, I think if you want to make it easy to choose predetermined gap settings specific to each matrix, (and that's a worthwhile usability improvement), the default should be NOT to set them. i.e. something like this?

```xml
<when value="BLOSUM62">
    <param name="gap_costs" type="select" label="Gap Costs">
        <option value="">Default values</option>
        <option value="-gapopen 11 -gapextend 2">Existence: 11  Extension: 2</option>
        <option value="-gapopen 10 -gapextend 2">Existence: 10  Extension: 2</option>
        ...
    </param>
</when>
```

**Comment by caleb-easterly** [comment]
My suspicion is that the gap costs are determined only by the matrix, but a quick scan through the source code didn't confirm or deny that. I like having the default penalties selected (rather than just "Default Options"), because a user could slightly change from the defaults (by knowing what the default was) without going to a different page to find out the defaults (which are actually in another page [here](https://www.ncbi.nlm.nih.gov/books/NBK379667/)). 

Do you know who could be contacted to request an update to the command line options page? I noticed that the `blastp-fast`, `blastx-fast`, and `tblastn-fast` tasks aren't documented there at all. If there's a link to it in the Galaxy wrapper it would be nice to have it as up-to-date and comprehensive as possible.

File: tools/ncbi_blast_plus/ncbi_blastx_wrapper.xml [hunk 8]
```diff
@@ -117,12 +123,20 @@ using the NCBI BLAST+ blastx command line tool.
...
 
 @OUTPUT_FORMAT@
 
 -------
 
+@CLI_OPTIONS@
+
+-------
+
+@SUBSTITUTION_MATRIX_PROTEIN@
```
Comment (nsoranzo) [inline tools/ncbi_blast_plus/ncbi_blastx_wrapper.xml]: You need to remove this now here and in ncbi_tblastn_wrapper.xml

**Comment by caleb-easterly** [comment]
Whoops, thanks for catching that.

File: tools/ncbi_blast_plus/ncbi_blastx_wrapper.xml (hunk 8 above)
Comment (caleb-easterly) [reply tools/ncbi_blast_plus/ncbi_blastx_wrapper.xml]: Whoops, thanks for catching that.

**Comment by caleb-easterly** [comment]
Is the Travis build failing because of something I did? It says that the SQL database is locked.

File: tools/ncbi_blast_plus/ncbi_macros.xml [hunk 9]
```diff
@@ -108,6 +111,160 @@
...
+                    <option value="-gapopen 14 -gapextend 1">Existence: 14 Extension: 1</option>
+                    <option value="-gapopen 10 -gapextend 1">Existence: 10  Extension: 1</option>
+                    <option value="-gapopen 9 -gapextend 1" selected="true">Existence: 9  Extension: 1 (default)</option>
+                    <option value="-gapopen 8 -gapextend 1">Existence: 8  Extension: 1</option>
+                </param>
+            </when>
+        </conditional>
+    </xml>
+    <!-- otherwise, the combined matrix/gapcosts macro is used -->
+    <!-- these gapopen and gapextend macros are only used in blastn -->
```
Comment (peterjc) [inline tools/ncbi_blast_plus/ncbi_macros.xml]: Are these two comment lines in the wrong order? And could they be one comment?

**Comment by peterjc** [comment]
Generally I would report BLAST issues via the NCBI helpdesk, who relay them to the developers. They can tell you the JIRA incident number (although you may need to ask for this explicitly) which is useful for cross-referencing.

Does my proposal yesterday https://github.com/peterjc/galaxy_blast/pull/96#issuecomment-330987131 not to add ``-gapopen`` and ``-gapextend`` to the command line by default (regardless of the matrix choice) seem sensible? Another argument in its favour is this is consistent with the current wrapper behaviour, which never sets these parameters.

**Comment by caleb-easterly** [comment]
Yes, I see the appeal of the consistency, and it would also be nice to have the defaults update automatically along with the version of blast+ (if they happen to change them), but I'm conflicted about not showing which gap costs are the default in the tool - I addressed that a little bit above. 

Also, I ran into the `--export-search-strategy` flag when I was looking at the command line options. Has the team considered adding support for that option? That would quiet some of my concerns about a user running a search without being able to go back later on and know what parameters were used - especially if they're not the type to look through source code. I don't know if that could be piped to stdout somehow, or just included as a separate history item. The ASN.1 format is surely not the prettiest, though.

**Comment by peterjc** [comment]
The NCBI themselves currently do not document the gap defaults in the command line tool, so I feel it is unreasonable for us as a wrapper to attempt to do so.

A separate pull request to add ``--export-search-strategy`` sounds interesting. Galaxy has ASN.1 as a datatype, but I don't know what the user could then do with it...

**Comment by caleb-easterly** [comment]
Sure, that makes sense. I'll add the empty defaults and push again tomorrow. 

The option to use the ASN.1 file to define the parameters for another search would probably be a little complicated (and not necessarily that useful), but at the very least it could document the search and contribute to reproducibility.

**Comment by peterjc** [comment]
Linting is failing, right now http://xfam.org/ is offline :(

**Comment by peterjc** [comment]
I'll need to review the changes again carefully, but this seems close to being ready 👍 

Would you be able to add a test or two using the new options? This should include an entry in https://github.com/peterjc/galaxy_blast/blob/master/tools/ncbi_blast_plus/update_test_files.sh which is used when updating BLAST itself (see separate issue #93 for the BLAST+ 2.6.0 work).

**Comment by caleb-easterly** [comment]
Yep, I'll add a couple of tests. Thanks!

**Comment by peterjc** [comment]
Merge conflict in ``tools/ncbi_blast_plus/ncbi_blastp_wrapper.xml`` from the addition of ``$adv_opts.use_sw_tback`` (and indentation), etc.

@caleb-easterly Are you familiar with ``git rebase`` and happy to rebase this to the current master branch?

**Comment by caleb-easterly** [comment]
The changelog looks good! I'm working on rebasing now - you said this will be a part of 0.2.01, but it looks like the wrapper version changed to 0.2.02. Should I keep the wrapper version at 0.2.01 in my fork, or update it?

**Comment by peterjc** [comment]
Sorry, typo - I think we'll do this as part of ``v0.2.02`` which thus far just does:

* Document the BLAST+ 2.5.0 change in the standard 12 column output from ``qseqid,sseqid,...`` to ``qacc,sacc,...`` instead.

Thanks for checking

**Comment by caleb-easterly** [comment]
Ok, I gave the rebase a shot. The history still shows a fair bit of fumbling around on my part, but I'm not really sure how to clean that up (or if I should).

**Comment by peterjc** [comment]
By eye the rebase looks good, waiting for TravisCI to confirm.

I'm tempted to squash the history down a bit - GitHub makes it easy to squash it to one single commit, if that's OK with you?

**Comment by caleb-easterly** [comment]
Sure, that's fine by me. I ended up reverting some of the things I did early on, so I think one commit would make it clearer what actually happened.

**Comment by caleb-easterly** [comment]
Looks like the build failed, maybe due to cancelling? I could also squash it if that's easier.

**Comment by peterjc** [comment]
Re-triggering the build seems to have fixed it. I'm currently off work sick, but will try to merge this soon after I'm back.

**Comment by caleb-easterly** [comment]
Sorry to hear that! Take care, and no hurry on the merge.

**Comment by peterjc** [comment]
Thanks all - that's merged and ought to automatically deploy to the Test Tool Shed shortly.

I think another release with this work prior to the BLAST+ 2.6.0 updates makes sense (#93)?

**Comment by caleb-easterly** [comment]
Thank you! :+1:
